core/src/main/scala/com/avast/clients/rabbitmq/DefaultRabbitMQStreamingConsumer.scala

```diff
@@ -0,0 +1,285 @@
+package com.avast.clients.rabbitmq
+
+import cats.effect.concurrent._
+import cats.effect.{Blocker, CancelToken, ConcurrentEffect, ContextShift, Effect, ExitCase, IO, Resource, Sync}
+import cats.syntax.all._
+import com.avast.bytes.Bytes
+import com.avast.clients.rabbitmq.DefaultRabbitMQConsumer.RepublishOriginalRoutingKeyHeaderName
+import com.avast.clients.rabbitmq.DefaultRabbitMQStreamingConsumer.DeliveryQueue
+import com.avast.clients.rabbitmq.api._
+import com.avast.metrics.scalaapi.Monitor
+import com.rabbitmq.client.{Delivery => _, _}
+import com.typesafe.scalalogging.StrictLogging
+import fs2.Stream
+import fs2.concurrent.Queue
+
+import scala.language.higherKinds
+import scala.util.control.NonFatal
+
+class DefaultRabbitMQStreamingConsumer[F[_]: ConcurrentEffect, A: DeliveryConverter] private (
+    name: String,
+    queueName: String,
+    initialConsumerTag: String,
+    connectionInfo: RabbitMQConnectionInfo,
+    consumerListener: ConsumerListener,
+    monitor: Monitor,
+    blocker: Blocker)(createQueue: F[DeliveryQueue[F, Bytes]], newChannel: F[ServerChannel])(implicit cs: ContextShift[F])
+    extends RabbitMQStreamingConsumer[F, A]
+    with StrictLogging {
+
+  private lazy val F: Sync[F] = Sync[F]
+
+  private lazy val streamFailureMeter = monitor.meter("streamFailures")
+
+  private lazy val recoveringMutex: Semaphore[F] = newMutexSemaphore()
+
+  private lazy val consumer = Ref.unsafe[F, Option[StreamingConsumer]](None)
+  private lazy val isClosed = Ref.unsafe[F, Boolean](false)
+  private lazy val isOk = Ref.unsafe[F, Boolean](false)
+  private lazy val tasks = Ref.unsafe[IO, Set[CancelToken[IO]]](Set.empty)
```

Comment:
perhaps here would better `Set[Fiber[F[_], A]]` since you can join on the Fiber and then there would be no need to call `unsafeRunSync` as well `Effect[F].toIO(action).runCancellable` down below.

Reply:
Actually Fiber would not work.